From aaaaff9e70de9da94608ca8927b18c286b5bfb10 Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megi@xff.cz>
Date: Sun, 19 Jul 2026 11:36:04 +0200
Subject: [PATCH] usb: musb: Fix pointer/integer casts in
 musb_rx_dma_iso_cppi41

The DMA address was computed via void pointer arithmetic with
truncating casts, warning on 64-bit builds and truncating DMA
addresses above 4 GiB.  channel_program() takes a dma_addr_t; compute
it as one.

Signed-off-by: Ondrej Jirman <megi@xff.cz>
---
 drivers/usb/musb/musb_host.c | 7 +++----
 1 file changed, 3 insertions(+), 4 deletions(-)

diff --git a/drivers/usb/musb/musb_host.c b/drivers/usb/musb/musb_host.c
index c250a7035527..cd08abcb9462 100644
--- a/drivers/usb/musb/musb_host.c
+++ b/drivers/usb/musb/musb_host.c
@@ -1464,12 +1464,11 @@ static int musb_rx_dma_iso_cppi41(struct dma_controller *dma,
 {
 	struct dma_channel *channel = hw_ep->rx_channel;
 	void __iomem *epio = hw_ep->regs;
-	dma_addr_t *buf;
+	dma_addr_t buf;
 	u32 length;
 	u16 val;
 
-	buf = (void *)urb->iso_frame_desc[qh->iso_idx].offset +
-		(u32)urb->transfer_dma;
+	buf = urb->transfer_dma + urb->iso_frame_desc[qh->iso_idx].offset;
 
 	length = urb->iso_frame_desc[qh->iso_idx].length;
 
@@ -1478,7 +1477,7 @@ static int musb_rx_dma_iso_cppi41(struct dma_controller *dma,
 	musb_writew(hw_ep->regs, MUSB_RXCSR, val);
 
 	return dma->channel_program(channel, qh->maxpacket, 0,
-				   (u32)buf, length);
+				   buf, length);
 }
 
 /* Host side RX (IN) using Mentor DMA works as follows:
-- 
2.43.0

